### ISSUE TITLE
Files created by the ODK should belong to the user who invoked the ODK

### ISSUE THREAD

**Issue by gouttegd** [INCATools/ontology-development-kit#682]
On GNU/Linux, files created by the ODK seeding script belong to the _root_ user.

This is because in most GNU/Linux installations, the Docker daemon itself runs under the identity of the super-user. On MacOS, the problem does not occur because the Docker daemon runs under the same identity as the user who started the Docker Desktop application.

It is possible to configure a GNU/Linux installation to have the Docker daemon run as a non-privileged user, but it’s non-standard and the behaviour of the ODK should not be dependent on the local Docker configuration — it should be up to the ODK to ensure that all the files it creates belong to the user who invoked the seeding script.

A solution to this problem had already been proposed in #225, but it involves the user building its own Docker image. We need a solution that works “out of the box” with the standard image.

**Comment by gouttegd** [comment]
So in theory, there is a very easy solution: just call `docker run` with `-u $(id -u):$(id -g)`, so that the process inside the Docker container runs under the user ID (`id -u`) and the group ID (`id -g`) of whoever is calling the wrapper script.

This works because, when specifying a user with a numerical ID (instead of a user name), Docker doesn’t need the user to already exist inside the Docker image. Likewise for the group.

**BUT…**

Doing so breaks our current method for providing Git with the informations it needs to create the initial commit when initialising a repository, because that method is dependent on the seeding process running under the `root` identity. Indeed, what we do is to bind the local user’s own `~/.gitconfig` file to `/root/.gitconfig` inside the container:

```sh
docker run -v $HOME/.gitconfig:/root/.gitconfig [...] /tools/odk.py seed "$@"
```

This obviously cannot work if the `odk.py` script runs under a non-root user.

Git needs the `.gitconfig` file to get the name and email of the committer when creating the initial commit after creating all the files. We need another way to provide Git with those informations.

**Comment by matentzn** [comment]
We could add these two pieces of information through the config, same way we add `main_branch`.. Are we sure these are the only reasons we bind this file? Dont you need it to "sign" the initial commit?

**Comment by gouttegd** [comment]
For now I am testing getting them in the wrapper script using `git config --get user.name` (and `--get user.email`), after updating the `odk.py` script to accept more options.

I do not _think_ we need this file for anything else, but we’ll see.

**Comment by gouttegd** [comment]
Reopening, since the solution implemented in #684 seems to be creating more problems than it solves.

**Comment by gouttegd** [comment]
# The problem

Processes within the ODK container run under the identity of the super-user. This is the default, and it allows to do virtually anything in the ODK workflows and in user-specific custom workflows (including installing extra packages, if necessary).

The problem is that the Docker daemon itself runs with _root_ privileges, then all files created as part of a workflow will belong to the super-user. This is not a problem on macOS because on macOS, the Docker daemon runs with the privileges of the local user. But it _is_ a problem on GNU/Linux, since unless configured otherwise, on GNU/Linux the Docker daemon does run with _root_ privileges.

# Possible solutions

There are two possible approaches to solve the issue.

## Not running as _root_ inside the container

This is what was implemented in #684. Processes inside the container run under the identity of the user who is invoking the `run.sh` script.

This causes two further problems:

* Because the ODK workflows no longer run with root privileges, any task that requires such privileges is no longer possible. The standard ODK workflows don’t have any such task, but there might be some custom workflows out there that do.
* Because we cannot know at build time what the UID of the calling user will be, we cannot create a home directory for that user inside the container, and we cannot populate `/etc/passwd` with an entry for that user.

The second point in turns cause several problems. For example, `python -m pip install --user`, which would normally install Python packages in the user’s home directory, will not work (since there is _no_ home directory). Some OAK commands won’t work either, because they expect to be able to cache some data somewhere under the user’s home directory.

We might work around the second point by using the [entrypoint](https://docs.docker.com/engine/reference/builder/#entrypoint) mechanism, with an entrypoint script that creates a home directory and add an entry in `/etc/passwd` before running the actual command.

The first point cannot be worked around however. If some custom workflows require root privileges, and we decide to support such workflows (I believe we should not, for what it’s worth), then we simply cannot avoid running the workflows as root.

## Running as root with user namespace re-mapping

The other possible solution is to keep running with root privileges inside the container. In that case, on GNU/Linux we need a way to remap the user namespace so that even if the processes inside the container run as root, from the host’s point of view they will appear to run with the local user’s privileges (so that any files created will belong to that user).

It’s unclear whether this can be done entirely automatically inside the `run.sh` script. It seems remapping users will require [some configuration](https://docs.docker.com/engine/security/userns-remap/) of the Docker daemon (possibly even some changes in the kernel configuration, as user namespaces are not always enabled by default on some distributions). In that case we should at least try to provide some documentation on how to do that.

**Comment by StroemPhi** [comment]
For me going into Docker rootless mode as described here https://docs.docker.com/engine/security/rootless/ seems to work without any problems.

**Comment by gouttegd** [comment]
Yes, I expect it would. But still, GNU/Linux users should not be required to alter the default configuration of Docker for the ODK to behave correctly.

**Comment by gouttegd** [comment]
Fixed with #769
